src/bufread.rs

```diff
@@ -80,44 +81,72 @@ impl<R: BufRead> BrotliEncoder<R> {
 }
 
 impl<R: BufRead> Read for BrotliEncoder<R> {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        if buf.len() == 0 {
-            return Ok(0)
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
+        if buf.is_empty() { return Ok(0) }
+        // If the compressor has failed at some point, this is set.
+        // Unfortunately we have no idea what status is in the compressor
+        // was in when it failed so we can't do anything except bail again.
+        if let Some(ref err) = self.err {
+            return Err(err.clone().into())
         }
 
-        match self.buf.read(buf) {
-            Ok(0) if self.done => return Ok(0),
-            Ok(0) => {
-                self.buf.get_mut().truncate(0);
-                self.buf.set_position(0);
-            }
-            other => return other,
+        if let Some(data) = self.data.take_output(Some(buf.len())) {
+            buf[..data.len()].copy_from_slice(data);
+            return Ok(data.len())
+        }
+
+        match self.done {
+            DoneStatus::Done => return Ok(0),
+            DoneStatus::Finishing => return tryfinish(self, buf),
+            DoneStatus::Processing => (),
         }
 
         loop {
-            assert!(self.cur < self.max);
-            let (amt_in, mut out) = {
-                let input = try!(self.obj.fill_buf());
-                let amt = cmp::min(input.len(), self.max - self.cur);
-                self.data.copy_input(&input[..amt]);
-                self.cur += amt;
-                (amt, try!(self.data.compress(amt == 0, false)))
-            };
-            self.cur = 0;
-            self.obj.consume(amt_in);
-            if amt_in == 0 {
-                self.done = true;
+            let amt_in;
+            let amt_out;
+            {
+                let input = &mut try!(self.obj.fill_buf());
+                let avail_in = input.len();
+                if avail_in == 0 {
+                    break
+                }
+                let mut output = &mut buf;
+                let avail_out = output.len();
+                if let Err(err) = self.data.compress(CompressOp::Process, input, output) {
+                    self.err = Some(err.clone().into());
```

Comment:
This pattern is present in a few places in the new code and is described by the comment at the top of the function:
```
        // If the compressor has failed at some point, this is set.
        // Unfortunately we have no idea what status is in the compressor
        // was in when it failed so we can't do anything except bail again.
```

In this specific case the compressor has failed and we have no idea what state it's in or whether we can continue using it (at least, the documentation says nothing so I don't really want to assume anything). To make sure this is 'retry safe', it needs to error forever now, and I figured repeating the original error would be reasonable.